src/pudl/extract/dbf.py

```diff
@@ -0,0 +1,452 @@
+
+    # TODO(rousik): table column map should be remapping short names (in dbf) to long names found in db schema (DBC).
+    # This is kind of annoying transformation but we can't do without it.
+
+    @lru_cache
+    def get_table_schema(self, table_name: str, year: int) -> DbfTableSchema:
+        """Returns TableSchema for a given table and a given year."""
+        table_columns = self.get_db_schema(year)[table_name]
+        dbf = self.get_table_dbf(table_name, year)
+        dbf_fields = [field for field in dbf.fields if field.name != "_NullFlags"]
+        if len(table_columns) != len(table_columns):
+            return ValueError(
+                f"Number of DBF fields in {table_name} does not match what was "
+                f"found in the DBC index file for {year}."
+            )
+        schema = DbfTableSchema(table_name)
+        for long_name, dbf_col in zip(table_columns, dbf_fields):
+            if long_name[:8] != dbf_col.name.lower()[:8]:
+                raise ValueError(
+                    f"DBF field name mismatch: {dbf_col.name} != {long_name}"
+                )
+            col_type = DBF_TYPES[dbf_col.type]
+            if col_type == sa.String:
+                col_type = sa.String(length=dbf_col.length)
+            schema.add_column(long_name, col_type, short_name=dbf_col.name)
+        return schema
+
+    def _load_single_year(self, table_name: str, year: int) -> pd.DataFrame:
+        """Returns dataframe that holds data for a single year for a given table.
+
+        Args:
+            table_name: name of the table.
+            year: year for which the data should be loaded.
+        """
+        sch = self.get_table_schema(table_name, year)
+        df = pd.DataFrame(iter(self.get_table_dbf(table_name, year)))
+        df = df.drop("_NullFlags", axis=1, errors="ignore").rename(
+            sch.get_column_rename_map(), axis=1
+        )
+        return df
+
+    def load_table_dfs(self, table_name: str, years: list[int]) -> pd.DataFrame | None:
+        """Returns the concatenation of the data for a given table and years.
+
+        Args:
+            table_name: name of the table to load.
+            years: list of years to load and concatenate.
+        """
+        yearly_dfs = []
+        for yr in years:
+            try:
+                yearly_dfs.append(self._load_single_year(table_name, yr))
+            except KeyError:
+                continue
+        if yearly_dfs:
+            return pd.concat(yearly_dfs, sort=True)
+        return None
+
+
+class FercDbfExtractor:
+    """Generalized class for loading data from foxpro databases into SQLAlchemy.
+
+    When subclassing from this generic extractor, one should implement dataset specific
+    logic in the following manner:
+
+    1. set DATABASE_NAME. This is going to be used as the file for the resulting sqlite
+    database.
+    2. Override get_datastore() method to return the right kind of dataset specific
+    datastore.
```

Comment:
It seems like there's not default implementation of `get_datastore()` to override. Is there supposed to be? Or is this a method that always has to be defined by child classes? It gets called in `__init__()`  By default should it just set  `self.datastore` to whatever datastore was passed in as an argument?

Reply:
I suppose we could make default implementation that throws NotImplemented exception. Given the renamings, this also needs to be renamed to `get_dbf_reader()`.

Reply:
Yeah that rename sounds good. So this is kind of an abstract class that lays out some infrastructure but has stuff that needs implementation.